README.md
A Dart linter enforcing certain naming conventions when dealing with methods that mutate variables.

There are 4 lints included in this package:

1. [mut_infect](#mut_infect)  
    ```dart
    void callAMut() {
        /* Calling a `mut` method should mark the caller as `Mut` */
        callBMut();
    }
    ```
1. [mut_out_of_scope](#mut_out_of_scope)
    ```dart
    void assignGlobalMut() {
        /* Assigning to a variable not owned by this scope should mark the caller as `Mut` */
        globalVar = 3;
    }
    ```
1. [mut_param](#mut_param)
    ```dart
    void changeIncomingParamterValue(AnObject aoMut) {
        /* Changing the parameter's properties should mark the parameter as `Mut` */
        aoMut.innerField = "hi";
    }
    ```
1. [unnecessary_mut_infect](#unnecessary_mut_infect)
    ```dart
    void funcDoesNotMut() {
        /* this function should not be marked with `Mut` */
    }
    ```


## Getting Started
Taken from [custom_lint](https://pub.dev/packages/custom_lint#enablingdisabling-and-configuring-lint):
* The application must contain an `analysis_options.yaml` with the following
```yaml
analyzer:
  plugins:
    - custom_lint
```
* The application also needs to add custom_lint and our package(s) as dev dependency in their application:
```yaml
# The pubspec.yaml of an application using our lints
name: example_app
environment:
  sdk: ">=2.16.0 <3.0.0"

dev_dependencies:
  custom_lint:
  lint_mut_infect:
```

### Disabling a lint
If you want to disable certain lints, add the following to the `analysis_options.yaml` file:
```yaml
lint_mut_infect:
  rules:
    - unnecessary_mut_infect: false # disable this rule
```

All rules are enabled by default.

# Lints
## mut_infect

Produces: Warning

Enforces infectious naming conventions on Method and Function declarations.   
Something that invokes a `Mut` element should also be called `Mut`.  


![Code demonstrating the `Mut Infect lint`, where a method that should be marked with Mut because it calls a method marked Mut](https://github.com/0xNF/dart_lint_mut_infect/blob/master/doc/readme/lint_mut_infect.png?raw=true)

This produces the following message: `'Mut' method invoked but not marked 'Mut'`

The suggested fix is to rename the calling function:

```dart
function markThisDummyMut() {
    dummyMut();
}
```

This way you can establish a chain of all functions that mutate variables.

### Exemptions
This lint is not applied when the element is:
1. Marked with `@override`
1. functions named `main`
1. A `setter`

## mut_out_of_scope
Produces: Warning

This lint checks that variables that are modified within a function are declared within that function, or are contained entirely within the lexical scope of the function in question. For instance:

![Code demonstrating the `Mut Out of Scope` lint, where a method is modifying a variable not declared in the lexical scope](https://github.com/0xNF/dart_lint_mut_infect/blob/master/doc/readme/lint_mut_out_of_scope.png?raw=true)

In this example, `globalScopeVar` is not declared within the `outOfScopeModifier` function, but is modified anyway. This produces a warning that the function should be marked with `Mut`.

This lint also understands locally defined functions, and won't cause undue warnings for strictly-local declarations:

![Code demonstrating that `Mut Out of Scope` doesn't apply to variables that have a perfectly captured lexical scope chain](https://github.com/0xNF/dart_lint_mut_infect/blob/master/doc/readme/lint_inner_funtions_not_included.png?raw=true)

In the above image, although `i` is modified by `inner()` , inner will not be marked as requiring Mut, because all declarations are local to the lexical scope of the top-level containing function.  

### Exemptions
This lint is not applied when the element is:
1. Marked with `@override`
1. functions named `main`
1. A `setter`

## mut_param
Produces: Error

This lint checks that any variable that is passed a parameter and is modified by the function is marked with `Mut`.   
This lint is stronger than the others because it is very important for a caller to know whether some object of theirs is going to be modified or not.

![Code demonstrating the `Mut Param` lint, where a parameter passed to the function is modified inside the function](https://github.com/0xNF/dart_lint_mut_infect/blob/master/doc/readme/lint_mut_unmarked_param.png?raw=true)



### Exemptions

* Parameters with Dart primitive types are exempt because they are passed-by-value. These include:
    - bool
    - int
    - double
    - num
    - String

* Simply reassigning the variable is exempt as well:
    ```dart
    void reassignment(AnObject ao) {
        ao = AnObject();
    }
    ```

## unnecessary_mut_infect
Produces: Warning

This lint checks that any functions or methods named `Mut` actually deserve to be marked. This lint can help keep refactors to your codebase from spiraling out of control with Muts everywhere. 

![Code demonstrating the `unnecessary_mut_infect` lint, where a function as been marked Mut that doesnt need it](https://github.com/0xNF/dart_lint_mut_infect/blob/master/doc/readme/lint_unnecessary_mut.png?raw=true)

### n.b.
Because this package isn't always aware of what constitutes mutating functionality or not, you should tell the the analyzer to ignore this lint when you need to.

For example:

```dart
// ignore: unnecessary_mut_infect
void reallyIsMut() {
    someCallYouCantRenameButIsDefinitelyAMutatingCall();
}
```


# Exemptions

Any function or method marked `@override`, or any function named `main` will not be flagged.  
This means you wont be overloaded with warnings when dealing with names you don't control.

# Limitations
This package doesn't integrate with `dart fix`, as we can't reliably get all symbol references for renaming. You are encouraged to use `F2` rename functionality on your own to add or remove `Mut` as appropriate. 



# Known Bugs
[#1](https://github.com/0xNF/dart_lint_mut_infect/issues/1)
Some functions with inner scope `Mut` calls mutating a variable owned entirely by the calling scope still require their parent scope to be marked as `Mut`, even though they're mutating fully-owned local variables.

# Debugging
To debug,

follow the steps at https://pub.dev/packages/custom_lint

1. Add `custom_lint` and `dart_lint_infect_mut` to Dev Dependencies of target project
2. Run `custom_lint --watch`
3. [Optional] add the dart_lint repo to the Workspace and set breakpoints within  


# Contributing to this plugin
This plugin uses the  [custom_lint](https://pub.dev/packages/custom_lint) package. To help with the underlying functionality, consider helping out that repo. To contribute to this repo specifically, visit https://github.com/0xNF/dart_lint_mut_infect

CHANGELOG.md
## 0.1.4

- Removed print statement in AST visitor

## 0.1.3

- Fixed typos in README

## 0.1.2

- Removed unused elements

## 0.1.1

- Marked many elements that were only intended for inner use private
- Added example project

## 0.1.0

- Initial version.

test/lint_mut_test.dart
// ignore_for_file: avoid_single_cascade_in_expression_statements, unnecessary_getters_setters, unused_local_variable, unused_element

int globalScopeVar = 0;

abstract class ForOveridding {
  void doSomeThing();
}

// FYI(nf, 09/08/23): Shouldn't be marked, because it is `main`
void main() {
  globalScopeVar = 3;
}

class BagB extends ForOveridding {
  String s = "";

// FYI(nf, 09/08/23): Shouldn't me marked, because it is `@override`
  @override
  void doSomeThing() {
    globalScopeVar = 3;
  }
}

class BagA {
  int i;
  List<String> strings;
  BagB b = BagB();

  BagA._({this.i = 0, this.strings = const <String>[]});

  factory BagA.sample(int i) {
    return BagA._(i: i);
  }

  // expect_lint: mut_out_of_scope
  void shouldMarkInBagA() {
    globalScopeVar = 1;
  }

  void create(String path) {}
}

/* This section shows that exemptions should be made for certain conditions */
abstract class MockWidget {
  int build();
}

class ImplMockWidget extends MockWidget {
  /* Shows that override method should not be considered */
  @override
  int build() {
    globalScopeVar = 2;
    return 0;
  }

  /* Shows that setters should not be considered */
  int get withSetter => globalScopeVar;
  set withSetter(int i) => globalScopeVar = i;
}

/* This section shows that reassignment to the variable at a depth of n=1 is fine, but reasignment to its inner items is not */
void reAssignDepth1(BagA bagADepth1) {
  bagADepth1 = BagA._();
}

// expect_lint: mut_param
void reAssignDepth2List(BagA bagADepth2List) {
  bagADepth2List.strings = <String>["dummy"];
}

// expect_lint: mut_param
void reAssignDepth2IntMut(BagA bagADepth2Int) {
  bagADepth2Int.i = 2;
}

// FYI(nf, 09/11/23): Cascaded inline assignments should not cause a lint warning
BagA createZipFile(String path) {
  final encoder = BagA._()..create(path);
  return encoder;
}

List<String> visible = [];
List<String> hidden = [];
List<String> makeVisible(String id) {
  final newVis = {...visible, id}.toList();
  final newHidden = [...hidden]..remove(id);

  return [...newVis, ...newHidden];
}

/// Special function to handle moving between LibCoordinator and LibGravioCloud
int dontThrowForComplexCascade() {
  final fromAuth1 = BagA._();
  final lc2auth = BagA.sample(fromAuth1.i)..i = 2;
  return 0;
}

int dontThrowForAssignmentOnConstructorCascade() {
  BagA._()..i = 12;
  return 0;
}

void dummyMut() {}

// expect_lint: mut_infect
void markThisDummy() {
  dummyMut();
}

void dontMarkThisDummyMut() {
  dummyMut();
}

// FYI(nf, 09/08/23): This should not be marked for any errors because it modifies nothing and calls no mut functions
void isInconsequential() {}

// expect_lint: mut_infect
void isNotMarked() {
  isMarkedMut();
}

// FYI(nf, 09/08/23): Should not be marked for errors because its marked mut
void isMarkedMut() {}

// FYI(nf, 09/08/23): Should be marked for Mut because it mutates an out of scope
// expect_lint: mut_out_of_scope
void mutsGlobalScope() {
  globalScopeVar = 2;
}

void mutsGlobalScopeMut() {
  globalScopeVar = 3;
}

// FYI(nf, 09/08/23): Should NOT be marked for errors because the variable is entirely within the function scope
void mutsInnerScope() {
  int i = 0;
  void inner() {
    i = 2;
  }

  print(i);

  inner();
}

// FYI(nf, 09/08/23): should be marked Mut for modifying outer scope
void mutsOuterScope() {
  int i = -1;

  // expect_lint: mut_out_of_scope
  void inner() {
    globalScopeVar = 4;
  }

  print(i);

  inner();
}

void mutsOuterScopeMut() {
  int i = 0;

  void innerMut() {
    globalScopeVar = 5;
  }

  print(i);

  innerMut();
}

// FYI(nf, 09/08/23): Should NOT be marked for errors because the variable is entirely within the function scope
void doesNotMutAnything() {
  void inner() {}

  inner();
}

// FYI(nf, 09/08/23): Should mark a as mut, simple
// expect_lint: mut_param
void markThisToo(BagA a) {
  a.i = 1;
}

void dontMarkThisToo(BagA aMut) {
  aMut.i = 1;
}

// expect_lint: mut_param
void markThisTooMut(BagA a) {
  a.i = 2;
}

void dontMarkThisTooMut(BagA aMut) {
  aMut.i = 2;
}

// expect_lint: mut_out_of_scope, mut_param
void markThisToo2(BagA a) {
  a..b.s = "lol";
}

// expect_lint: mut_out_of_scope
void dontMarkThisToo2(BagA aMut) {
  aMut..b.s = "lol";
}

// expect_lint: mut_param
void markThisToo3Mut(BagA b) {
  b
    ..b.s = "lmao"
    ..i = 3;
}

void dontMarkThisToo3Mut(BagA bMut) {
  bMut
    ..b.s = "lmao"
    ..i = 2;
}

/* The following items won't be marked for MutParam because they are Dart primitive types, which are passed by value */
void dontMarkRequiredPrimitiveInt(int i, {required int i2}) {
  i2 = 3;
}

void dontMarkDefaultPrimitiveInt(int i, {int i3 = 45}) {
  i3 = 4;
}

void dontMarkPositionalPrimitiveInt(int i, [int i4 = 0]) {
  i4 = 4;
}

void dontMarkRequiredPrimitiveDouble(int i, {required double i5}) {
  i5 = 5.0;
}

void dontMarkDefaultPrimitiveDouble(int i, {double i6 = 45.0}) {
  i6 = 6.0;
}

void dontMarkPositionalPrimitiveDouble(int i, [double i7 = 0.0]) {
  i7 = 7.0;
}

void dontMarkRequiredPrimitiveNum(int i, {required num i8}) {
  i8 = 5.0;
}

void dontMarkDefaultPrimitiveNum(int i, {num i9 = 45.0}) {
  i9 = 6.0;
}

void dontMarkPositionalPrimitiveNum(int i, [num i10 = 0.0]) {
  i10 = 7.0;
}

void dontMarkRequiredPrimitiveBool(int i, {required bool bool0}) {
  bool0 = false;
}

void dontMarkDefaultPrimitiveBool(int i, {bool bool1 = true}) {
  bool1 = false;
}

void dontMarkPositionalPrimitiveBool(int i, [bool bool2 = true]) {
  bool2 = false;
}

void dontMarkRequiredPrimitiveString(int i, {required String str0}) {
  str0 = "ay";
}

void dontMarkDefaultPrimitiveString(int i, {String str1 = "abc"}) {
  str1 = "lmao";
}

void dontMarkPositionalPrimitiveString(int i, [String str2 = "xyz"]) {
  str2 = "lol";
}


test/lint_mut_infect_test.dart
import 'package:lint_mut_infect/lint_mut_infect.dart';
import 'package:test/test.dart';

void isInconsequential() {}

// expect_lint: mut_infect
void isNotMarked() {
  isMarkedMut();
}

void isMarkedMut() {}


example/test/lint_mut_infect_example_test.dart
// import 'package:lint_mut_testy/lint_mut_testy.dart';
// import 'package:test/test.dart';

// void main() {
//   test('calculate', () {
//     expect(calculate(), 42);
//   });
// }


example/lib/lint_mut_infect_example.dart
// ignore_for_file: avoid_single_cascade_in_expression_statements, unnecessary_getters_setters, unused_local_variable, unused_element

int globalScopeVar = 0;

abstract class ForOveridding {
  void doSomeThing();
}

// Shouldn't be marked, because it is `main`
void main() {
  globalScopeVar = 3;
}

class BagB extends ForOveridding {
  String s = "";

// Shouldn't me marked, because it is `@override`
  @override
  void doSomeThing() {
    globalScopeVar = 3;
  }

  // expect_lint: unnecessary_mut_infect
  void incorrectMut() {}

  // expect_lint: unnecessary_mut_infect
  void incorrect2Mut(BagA functionParamIsNotMut) {
    print(functionParamIsNotMut);
  }
}

class BagA {
  int i;
  List<String> strings;
  BagB b = BagB();

  BagA._({this.i = 0, this.strings = const <String>[]});

  factory BagA.sample(int i) {
    return BagA._(i: i);
  }

  // expect_lint: mut_out_of_scope
  void shouldMarkInBagA() {
    globalScopeVar = 1;
  }

  void create(String path) {}
}

/* This section shows that items that don't do any mutating should not be marked as mut */
// expect_lint: unnecessary_mut_infect
void incorrectlyMarkedMut() {}

// expect_lint: unnecessary_mut_infect
void incorrectlyMarked2Mut(BagA methodParamIsNotMut) {
  print(methodParamIsNotMut);
}

/* This section shows that exemptions should be made for certain conditions */
abstract class MockWidget {
  int build();
}

class ImplMockWidget extends MockWidget {
  /* Shows that override method should not be considered */
  @override
  int build() {
    globalScopeVar = 2;
    return 0;
  }

  /* Shows that setters should not be considered */
  int get withSetter => globalScopeVar;
  set withSetter(int i) => globalScopeVar = i;
}

/* This section shows that reassignment to the variable at a depth of n=1 is fine, but reasignment to its inner items is not */
void reAssignDepth1(BagA bagADepth1) {
  bagADepth1 = BagA._();
}

// expect_lint: mut_param
void reAssignDepth2List(BagA bagADepth2List) {
  bagADepth2List.strings = <String>["dummy"];
}

// expect_lint: mut_param
void reAssignDepth2IntMut(BagA bagADepth2Int) {
  bagADepth2Int.i = 2;
}

BagA createZipFile(String path) {
  final encoder = BagA._()..create(path);
  return encoder;
}

List<String> visible = [];
List<String> hidden = [];
List<String> makeVisible(String id) {
  final newVis = {...visible, id}.toList();
  final newHidden = [...hidden]..remove(id);

  return [...newVis, ...newHidden];
}

/// Special function to handle moving between LibCoordinator and LibGravioCloud
int dontThrowForComplexCascade() {
  final fromAuth1 = BagA._();
  final lc2auth = BagA.sample(fromAuth1.i)..i = 2;
  return 0;
}

int dontThrowForAssignmentOnConstructorCascade() {
  BagA._()..i = 12;
  return 0;
}

// expect_lint: unnecessary_mut_infect
void dummyMut() {}

// expect_lint: mut_infect
void markThisDummy() {
  dummyMut();
}

void dontMarkThisDummyMut() {
  dummyMut();
}

// This should not be marked for any errors because it modifies nothing and calls no mut functions
void isInconsequential() {}

// expect_lint: mut_infect
void isNotMarked() {
  isMarkedMut();
}

// Should not be marked for errors because its marked mut
// expect_lint: unnecessary_mut_infect
void isMarkedMut() {}

//  Should be marked for Mut because it mutates an out of scope
// expect_lint: mut_out_of_scope
void mutsGlobalScope() {
  globalScopeVar = 2;
}

void mutsGlobalScopeMut() {
  globalScopeVar = 3;
}

// Should NOT be marked for errors because the variable is entirely within the function scope
void mutsInnerScope() {
  int i = 0;
  void inner() {
    i = 2;
  }

  print(i);

  inner();
}

// should be marked Mut for modifying outer scope
void mutsOuterScope() {
  int i = -1;

  // expect_lint: mut_out_of_scope
  void inner() {
    globalScopeVar = 4;
  }

  print(i);

  inner();
}

void mutsOuterScopeMut() {
  int i = 0;

  void innerMut() {
    globalScopeVar = 5;
  }

  print(i);

  innerMut();
}

// Should NOT be marked for errors because the variable is entirely within the function scope
void doesNotMutAnything() {
  void inner() {}

  inner();
}

// expect_lint: mut_param
void markThisToo(BagA a) {
  a.i = 1;
}

void dontMarkThisToo(BagA aMut) {
  aMut.i = 1;
}

// expect_lint: mut_param
void markThisTooMut(BagA a) {
  a.i = 2;
}

void dontMarkThisTooMut(BagA aMut) {
  aMut.i = 2;
}

// expect_lint: mut_out_of_scope, mut_param
void markThisToo2(BagA a) {
  a..b.s = "lol";
}

// expect_lint: mut_out_of_scope
void dontMarkThisToo2(BagA aMut) {
  aMut..b.s = "lol";
}

// expect_lint: mut_param
void markThisToo3Mut(BagA b) {
  b
    ..b.s = "lmao"
    ..i = 3;
}

void dontMarkThisToo3Mut(BagA bMut) {
  bMut
    ..b.s = "lmao"
    ..i = 2;
}

/* The following items won't be marked for MutParam because they are Dart primitive types, which are passed by value */
void dontMarkRequiredPrimitiveInt(int i, {required int i2}) {
  i2 = 3;
}

void dontMarkDefaultPrimitiveInt(int i, {int i3 = 45}) {
  i3 = 4;
}

void dontMarkPositionalPrimitiveInt(int i, [int i4 = 0]) {
  i4 = 4;
}

void dontMarkRequiredPrimitiveDouble(int i, {required double i5}) {
  i5 = 5.0;
}

void dontMarkDefaultPrimitiveDouble(int i, {double i6 = 45.0}) {
  i6 = 6.0;
}

void dontMarkPositionalPrimitiveDouble(int i, [double i7 = 0.0]) {
  i7 = 7.0;
}

void dontMarkRequiredPrimitiveNum(int i, {required num i8}) {
  i8 = 5.0;
}

void dontMarkDefaultPrimitiveNum(int i, {num i9 = 45.0}) {
  i9 = 6.0;
}

void dontMarkPositionalPrimitiveNum(int i, [num i10 = 0.0]) {
  i10 = 7.0;
}

void dontMarkRequiredPrimitiveBool(int i, {required bool bool0}) {
  bool0 = false;
}

void dontMarkDefaultPrimitiveBool(int i, {bool bool1 = true}) {
  bool1 = false;
}

void dontMarkPositionalPrimitiveBool(int i, [bool bool2 = true]) {
  bool2 = false;
}

void dontMarkRequiredPrimitiveString(int i, {required String str0}) {
  str0 = "ay";
}

void dontMarkDefaultPrimitiveString(int i, {String str1 = "abc"}) {
  str1 = "lmao";
}

void dontMarkPositionalPrimitiveString(int i, [String str2 = "xyz"]) {
  str2 = "lol";
}
